RavenDB-17243: Support for empty at the index entry writer level.

redknightlois opened the pull request.

redknightlois committed:
RavenDB-17243: Support for empty at the index entry writer level.

maciejaszyk commented:
The RavenDB part of this PR is missing.

redknightlois commented:
What you mean? What I am missing?

redknightlois commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentConverterBase.cs:
```diff
@@ -304,7 +304,7 @@ public IndexFieldsMapping GetKnownFields()
                 throw new NotSupportedException("Boosting in index is not supported by Corax. You can do it during querying or change index type into Lucene.");
             case ValueType.EmptyString:
-                scope.Write(field.Id, EmptyString.Span, ref entryWriter);
+                scope.Write(field.Id, ReadOnlySpan<byte>.Empty, ref entryWriter);
```
@maciejaszyk Found it.

arekpalinski commented:
test this please

arekpalinski commented:
On macOS we got (https://github.com/ravendb/ravendb/runs/6628780917?check_suite_focus=true):

```
Failed FastTests.Corax.PrimitivesTests.OutputCompatibilityForVectorizedSingleRun(seed: 125, ai: 1, bi: 24, ss: 1) [< 1 ms]
  Error Message:
   System.PlatformNotSupportedException : Operation is not supported on this platform.
  Stack Trace:
     at System.Runtime.Intrinsics.X86.Avx2.CompareEqual(Vector256`1 left, Vector256`1 right)
   at Corax.Queries.MergeHelper.AndVectorized(Span`1 dst, Span`1 left, Span`1 right) in /Users/runner/work/ravendb/ravendb/src/Corax/Queries/MergeHelper.cs:line 145
   at FastTests.Corax.PrimitivesTests.OutputCompatibilityForVectorizedSingleRun(Int32 seed, Int32 ai, Int32 bi, Int32 ss) in /Users/runner/work/ravendb/ravendb/test/FastTests/Corax/Primitives.cs:line 141
   ```

arekpalinski commented:
Also here we got failure in a test with `AndIn` (https://github.com/ravendb/ravendb/runs/6628781130?check_suite_focus=true):

```
Failed FastTests.Corax.IndexSearcherTest.AndInStatementAndWhitespaceTokenizer(setSize: 100000, stackSize: 2046) [5 s]
  Error Message:
   Assert.Equal() Failure
Expected: 33333
Actual:   33306
  Stack Trace:
     at FastTests.Corax.IndexSearcherTest.AndInStatementAndWhitespaceTokenizer(Int32 setSize, Int32 stackSize) in D:\a\ravendb\ravendb\test\FastTests\Corax\IndexSearcher.cs:line 1470

arekpalinski commented:
test this please

arekpalinski commented:
Also please check `SlowTests.Issues.Q14235692.SlowTests.Issues.Q14235692.Empty_Strings_Can_Be_Used_In_Where_In_With_Other_Data(options: DatabaseMode = Single , SearchEngineMode = Corax)` 

and

`[SlowTests.Bugs.Queries.QueryingOnEmptyString.SlowTests.Bugs.Queries.QueryingOnEmptyString.CanFindByemptyStringMatch(options: DatabaseMode = Single , SearchEngineMode = Corax)](https://jenkins.hibernatingrhinos.com/job/PR_Tests/2046/testReport/junit/SlowTests.Bugs.Queries/QueryingOnEmptyString/SlowTests_Bugs_Queries_QueryingOnEmptyString_CanFindByemptyStringMatch_options___DatabaseMode___Single___SearchEngineMode___Corax_/)` 
@maciejaszyk can you take a look as well?

redknightlois commented:
@arekpalinski Is this the first time you see these MacOS error and in `AndIn`?

EDIT: Just pushed some more stringent initialization, to see how it behaves on the test server.

redknightlois committed:
RavenDB-17243: Do not call `AndVectorized` where Avx2 is not available.

redknightlois committed:
RavenDB-17243: Solving the heisenbug in AndIn

The heisenbug was probably caused by the following sequence:
- A value was stored in `_doubleOffset` and then reclaimed on the stack
- Following call would initialize it with an empty list, and given we are not initializing we will use the previous value in the stack
- When Sequence gets called, now IsTuple would look for the IsTuple and if the value was not zero would return.

redknightlois committed:
RavenDB-18708: It may happen that when matches is smaller than N we could go over the boundary

redknightlois committed:
RavenDB-17243: Fixed the case of simple empty strings.
